File: apps/master-app/src/bridge/discord-setup.ts
import {
  discordGuildMemberOptionSchema,
  discordGuildOptionSchema,
  discordOauthStatusSchema,
  discordVoiceChannelOptionSchema,
  type DiscordGuildMemberOption,
  type DiscordGuildOption,
  type DiscordOauthStatus,
  type DiscordVoiceChannelOption,
} from '@amber/shared';
import { invoke } from '@tauri-apps/api/core';

export type {
  DiscordGuildMemberOption,
  DiscordGuildOption,
  DiscordOauthStatus,
  DiscordVoiceChannelOption,
};

export async function discordOauthStart(): Promise<void> {
  return invoke<void>('discord_oauth_start');
}

export async function discordOauthClear(): Promise<void> {
  return invoke<void>('discord_oauth_clear');
}

export async function discordOauthStatus(): Promise<DiscordOauthStatus> {
  const raw = await invoke<unknown>('discord_oauth_status');
  return discordOauthStatusSchema.parse(raw);
}

export async function discordOauthLogout(): Promise<void> {
  return invoke<void>('discord_oauth_logout');
}

export async function discordEnsureUserOauth(): Promise<void> {
  return invoke<void>('discord_ensure_user_oauth');
}

let adminGuildsInflight: Promise<DiscordGuildOption[]> | null = null;

export async function discordListAdminGuilds(): Promise<DiscordGuildOption[]> {
  if (adminGuildsInflight) return adminGuildsInflight;

  adminGuildsInflight = invoke<unknown[]>('discord_list_admin_guilds')
    .then((raw) => raw.map((row) => discordGuildOptionSchema.parse(row)))
    .finally(() => {
      adminGuildsInflight = null;
    });

  return adminGuildsInflight;
}

export async function discordListGuildMembers(
  guildId: string,
): Promise<DiscordGuildMemberOption[]> {
  const raw = await invoke<unknown[]>('discord_list_guild_members', { guildId });
  return raw.map((row) => discordGuildMemberOptionSchema.parse(row));
}

export async function discordSearchGuildMembers(
  guildId: string,
  query: string,
): Promise<DiscordGuildMemberOption[]> {
  const raw = await invoke<unknown[]>('discord_search_guild_members', { guildId, query });
  return raw.map((row) => discordGuildMemberOptionSchema.parse(row));
}

export async function discordOpenBotInvite(guildId: string): Promise<void> {
  return invoke<void>('discord_open_bot_invite', { guildId });
}

export async function discordIsBotInGuild(guildId: string): Promise<boolean> {
  return invoke<boolean>('discord_is_bot_in_guild', { guildId });
}

export async function discordListVoiceChannels(
  guildId: string,
): Promise<DiscordVoiceChannelOption[]> {
  const raw = await invoke<unknown[]>('discord_list_voice_channels', { guildId });
  return raw.map((row) => discordVoiceChannelOptionSchema.parse(row));
}

export async function discordParseBotApplicationId(token: string): Promise<string> {
  return invoke<string>('discord_parse_bot_application_id', { token });
}

File: apps/master-app/src-tauri/src/commands/discord_setup.rs
use std::sync::Arc;

use tauri::{AppHandle, State};

use crate::error::AppError;
use crate::services::discord_setup::{
    self, application_id_from_bot_token, DiscordGuildMemberOption, DiscordOAuthState,
    DiscordGuildOption, DiscordOauthStatus, DiscordVoiceChannelOption,
};

#[tauri::command]
pub async fn discord_oauth_start(
    app: AppHandle,
    oauth: State<'_, Arc<DiscordOAuthState>>,
) -> Result<(), AppError> {
    discord_setup::oauth_start(&app, oauth.inner()).await
}

#[tauri::command]
pub fn discord_oauth_clear(oauth: State<'_, Arc<DiscordOAuthState>>) -> Result<(), AppError> {
    discord_setup::oauth_clear(oauth.inner());
    Ok(())
}

#[tauri::command]
pub fn discord_oauth_status() -> Result<DiscordOauthStatus, AppError> {
    discord_setup::oauth_status()
}

#[tauri::command]
pub fn discord_oauth_logout(oauth: State<'_, Arc<DiscordOAuthState>>) -> Result<(), AppError> {
    discord_setup::oauth_logout(oauth.inner())
}

#[tauri::command]
pub async fn discord_ensure_user_oauth(
    app: AppHandle,
    oauth: State<'_, Arc<DiscordOAuthState>>,
) -> Result<(), AppError> {
    discord_setup::ensure_user_oauth(&app, oauth.inner()).await
}

/// Valid Discord user access token for HTTP API master session JWT exchange.
#[tauri::command]
pub async fn discord_user_access_token(
    app: AppHandle,
    oauth: State<'_, Arc<DiscordOAuthState>>,
) -> Result<String, AppError> {
    discord_setup::ensure_user_access_token(&app, oauth.inner()).await
}

#[tauri::command]
pub async fn discord_list_guild_members(
    app: AppHandle,
    guild_id: String,
) -> Result<Vec<DiscordGuildMemberOption>, AppError> {
    discord_setup::list_guild_members(&app, &guild_id).await
}

#[tauri::command]
pub async fn discord_search_guild_members(
    app: AppHandle,
    guild_id: String,
    query: String,
) -> Result<Vec<DiscordGuildMemberOption>, AppError> {
    discord_setup::search_guild_members(&app, &guild_id, &query).await
}

#[tauri::command]
pub async fn discord_list_admin_guilds(
    app: AppHandle,
    oauth: State<'_, Arc<DiscordOAuthState>>,
) -> Result<Vec<DiscordGuildOption>, AppError> {
    discord_setup::list_admin_guilds_with_bot(&app, oauth.inner()).await
}

#[tauri::command]
pub fn discord_open_bot_invite(app: AppHandle, guild_id: String) -> Result<(), AppError> {
    discord_setup::open_bot_invite(&app, &guild_id)
}

#[tauri::command]
pub async fn discord_is_bot_in_guild(app: AppHandle, guild_id: String) -> Result<bool, AppError> {
    discord_setup::is_bot_in_guild(&app, &guild_id).await
}

#[tauri::command]
pub async fn discord_list_voice_channels(
    app: AppHandle,
    guild_id: String,
) -> Result<Vec<DiscordVoiceChannelOption>, AppError> {
    discord_setup::list_voice_channels(&app, &guild_id).await
}

#[tauri::command]
pub fn discord_parse_bot_application_id(token: String) -> Result<String, AppError> {
    application_id_from_bot_token(&token)
}
